src/btree/bt_cursor.c

```diff
@@ -2296,14 +2302,56 @@ __wt_btcur_bounds_position(
 
     if (CUR2BT(cursor)->type == BTREE_ROW) {
         WT_RET(__cursor_row_search(cbt, false, NULL, NULL));
-        /*
-         * If there's an exact match, the row-store search function built the key in the cursor's
-         * temporary buffer.
-         */
-        WT_RET(__wt_cursor_valid(cbt, (cbt->compare == 0 ? cbt->tmp : NULL), WT_RECNO_OOB, &valid));
+        if (cbt->compare == 0) {
+            /*
+             * If our compare value is 0 we found the exact key we were looking for, if the bound we
+             * are searching for is inclusive we can return it without walking.
+             */
+            if (F_ISSET(
+                  cursor, next ? WT_CURSTD_BOUND_LOWER_INCLUSIVE : WT_CURSTD_BOUND_UPPER_INCLUSIVE))
+                WT_RET(__wt_cursor_valid(cbt, cbt->tmp, WT_RECNO_OOB, &valid, false));
+            else
+                *need_walkp = true;
+        } else if (cbt->compare > 0) {
```

Comment:
what do you think combining the if statements as (cbt->compare > 0 && next) || (cbt->compare < 0 && !next)? We can reduce down to one else case too, and have less need for comments.

Reply:
Sounds good to me.

Reply:
[94a0b7c](https://github.com/wiredtiger/wiredtiger/pull/8309/commits/94a0b7c7710e831d2048fa14103fc83ecd152ffb)